### ISSUE TITLE
Struggling with the understanding of .Include() and why it produces Inner joins (and how to change it to left outer joins)

### ISSUE THREAD

**Issue by MontyGvMC** [dotnet/efcore#9051]
[SelectTest.zip](https://github.com/aspnet/EntityFramework/files/1119415/SelectTest.zip)

In a .NET Core Web API I have a controller which simply should return a list of entities with optional (foreign key is nullable) referenced entities which can be filtered by one criteria and paginated.

According to the famous Blog example I simply used the equivalent to
```C#
dbContext.Blogs.Include(b => b.Posts)
```

In my example
```C#
dbContext.Employee
.Include(ee => ee.Department)
.Include(ee => ee.SalaryBoundary)
.Include(ee => ee.Project)
.Include(ee => ee.EmploymentStatus)
```

Don't mind the strange table and column names, they come from obfuscating my real world names.

The problem is that it always generates an inner join and since one of the referenced elements is null the result is always an empty list.
And I can't figure out how I can bring EF core to generate left (outer) joins. And still don't get it why the default behavior for such an example seem to be inner join.

I attached an example project. The (local) database (including data) can be created with the Create-DB.sql script and the connection string can be configured in appsettings.json. The example project provides a setup for testing the problem with swagger and see the generated log files in the logs directory.

Please can anyone show me how to change the query to left outer joins so the list returns all eight entries with the view elements attached and null if foreign key is null.

The full controller causing the problems is

```C#
// the controller
    [Route("api/v1/[controller]")]
    public class EmployeeController : Controller
    {

        private readonly DatabaseContext _dbContext;
        private readonly ILogger<EmployeeController> _logger;

        /// <summary>
        /// Constructs a new instance of EmployeeController
        /// </summary>
        /// <param name="dbContext">the database context to be used</param>
        /// <param name="logger">the logger to be used</param>
        public EmployeeController(DatabaseContext dbContext, ILogger<EmployeeController> logger)
        {
            _dbContext = dbContext;
            _logger = logger;
            _logger.LogDebug("{0} > new instance created", GetType().Name);
        }

        /// <summary>
        /// Returns a list of Employees
        /// </summary>
        /// <param name="hideInactive">optional if inactive employees  should be excluded or not</param>
        /// <returns></returns>
        /// <response code="200">Ok</response>
        /// <response code="400">Bad Request</response>
        [HttpGet(Name = "ListEmployees")]
        [Produces("application/json")]
        [ProducesResponseType(typeof(ICollection<EmployeeView>), 200)]
        [ProducesResponseType(typeof(Microsoft.AspNetCore.Mvc.ModelBinding.ModelStateDictionary), 400)]
        [SwaggerOperation("ListEmployees")]
        
        public IActionResult ListEmployees([FromQuery] bool hideInactive = false)
        {
            _logger.LogInformation("{0}.ListEmployees(hideInactive: {1})", GetType().Name, hideInactive);

            if (!ModelState.IsValid)
            {
                _logger.LogWarning("{0}.ListEmployees(hideInactive: {1}) > bad request due to invalid ModelState: {2}", GetType().Name, hideInactive, JsonConvert.SerializeObject(ModelState));
                return BadRequest(ModelState);
            }

            var query = _dbContext.Employee
                .Include(ee => ee.Department)
                .Include(ee => ee.SalaryBoundary)
                .Include(ee => ee.Project)
                .Include(ee => ee.EmploymentStatus)
                .AsNoTracking()
                .AsQueryable();

            if (hideInactive)
            {
                query = query.Where(ee => ee.IsActive == true);
            }

            query = query.OrderBy(ee => ee.Id);


            Pagination pagintation = Pagination.FromRequest(Request);

            if (pagintation.IsEnabled)
            {
                pagintation.TotalElements = query.Count();

                query = query.Skip(pagintation.SkipItems).Take(pagintation.ItemsPerPage);

                Pagination.ToResponse(Response, pagintation);
            }

            List<EmployeeView> result = query
                                        .Select(ee => new EmployeeView()
                                            {
                                                Id = ee.Id,
                                                Name = ee.Name,
                                                IsActive = ee.IsActive,

                                                Project = (ee.Project == null) ? null :
                                                    new ProjectView()
                                                    {
                                                        Name = ee.Project.Name,
                                                        ClientCount = ee.Project.ClientCount,
                                                        PriorityLevel = ee.Project.PriorityLevel,
                                                        RiskLevel = ee.Project.RiskLevel
                                                    },

                                                EmploymentStatus = (ee.EmploymentStatus == null) ? null :
                                                    new EmploymentStatusView()
                                                    {
                                                        Name = ee.EmploymentStatus.Name
                                                    },

                                                SalaryBoundary = (ee.SalaryBoundary == null) ? null :
                                                    new SalaryBoundaryView()
                                                    {
                                                        SalaryLimit = ee.SalaryBoundary.SalaryLimit
                                                    },

                                                Department = (ee.Department == null) ? null :
                                                    new DepartmentView()
                                                    {
                                                        Name = ee.Department.Name,
                                                        EmployeeCount = ee.Department.EmployeeCount,
                                                        IsSuccessfull = ee.Department.IsSuccessfull,
                                                        AverageIncome = ee.Department.AverageIncome
                                                    }
                                            }
                                        )
                                        .ToList();


            return Ok(result);

        }
       
        /// <summary>
        /// Returns the Employee with the given id
        /// </summary>
        /// <param name="id">obligatory id of the Employee</param>
        /// <returns></returns>
        /// <response code="200">Ok</response>
        /// <response code="400">Bad Request</response>
        /// <response code="404">Not Found</response>
        [HttpGet("{id:int}", Name = "GetEmployeeById")]
        [Produces("application/json")]
        [ProducesResponseType(typeof(EmployeeView), 200)]
        [ProducesResponseType(typeof(ICollection<NotFoundInfo>), 404)]
        [ProducesResponseType(typeof(Microsoft.AspNetCore.Mvc.ModelBinding.ModelStateDictionary), 400)]
        [SwaggerOperation("GetEmployeeById")]
        public IActionResult GetEmployeeById(int id)
        {
            _logger.LogInformation("{0}.GetEmployeeById(id: {1})", GetType().Name, id);

            if (!ModelState.IsValid)
            {
                _logger.LogWarning("{0}.GetEmployeeById(id: {1}) > bad request due to invalid ModelState: {2}", GetType().Name, id, JsonConvert.SerializeObject(ModelState));
                return BadRequest(ModelState);
            }

           EmployeeView result = _dbContext.Employee
                .Include(mp => mp.Department)
                .Include(mp => mp.SalaryBoundary)
                .Include(mp => mp.Project)
                .Include(mp => mp.EmploymentStatus)
                .Where(mpf => mpf.Id == id)
                .AsNoTracking()
                .Select(ee => new EmployeeView()
                    {
                        Id = ee.Id,
                        Name = ee.Name,
                        IsActive = ee.IsActive,

                        Project = (ee.Project == null) ? null :
                                                    new ProjectView()
                                                    {
                                                        Name = ee.Project.Name,
                                                        ClientCount = ee.Project.ClientCount,
                                                        PriorityLevel = ee.Project.PriorityLevel,
                                                        RiskLevel = ee.Project.RiskLevel
                                                    },

                        EmploymentStatus = (ee.EmploymentStatus == null) ? null :
                                                    new EmploymentStatusView()
                                                    {
                                                        Name = ee.EmploymentStatus.Name
                                                    },

                        SalaryBoundary = (ee.SalaryBoundary == null) ? null :
                                                    new SalaryBoundaryView()
                                                    {
                                                        SalaryLimit = ee.SalaryBoundary.SalaryLimit
                                                    },

                        Department = (ee.Department == null) ? null :
                                                    new DepartmentView()
                                                    {
                                                        Name = ee.Department.Name,
                                                        EmployeeCount = ee.Department.EmployeeCount,
                                                        IsSuccessfull = ee.Department.IsSuccessfull,
                                                        AverageIncome = ee.Department.AverageIncome
                                                    }
                    }
                )
                .SingleOrDefault();

            if (result == null)
            {
                return NotFound(new NotFoundInfo("id", id).AsCollection());
            }

            return Ok(result);

        }
    }

}
```

### Further technical details
EF Core version: 1.1.0
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 7
IDE: Visual Studio 2015
IIS Express

**Comment by ajcvickers** [comment]
@MontyGvMC In general, a required relationship (that is, a non-nullable FK) will result in an inner join, because it is not possible for the dependent table to have any null FK values. Looking at your model, it seems some of the relationships are required, so inner join will be used.

Also note that it looks like none of the Includes are being used by the query. You might want to take a look at this announcement: https://github.com/aspnet/Announcements/issues/256

**Comment by MontyGvMC** [comment]
Thank you for taking time to answer. 

Unfortunatly I still have no idea how the code is intended to be so it returns all eight entries (creating left outer join).

**Comment by divega** [comment]
@MontyGvMC @ajcvickers The query ends up using INNER JOIN instead of LEFT OUTER JOIN because somehow the model is constructed with required relationships, despite the fact that all but one FKs on `Employee` are nullable. 

I was able to workaround this by explicitly calling `IsRequired(false)` on the relationships, e.g.:

``` C#
    entity.HasOne(d => d.Department)
        .WithMany(p => p.Employee)
        .HasForeignKey(d => d.DepartmentId)
        .IsRequired(false)
        .OnDelete(DeleteBehavior.Restrict)
        .HasConstraintName("FK_Employee_Department");

```

But I suspect there is either a bug or I am missing some detail in the model configuration.

**Comment by divega** [comment]
What I was missing are all the calls to `IsRequired()` on the FK properties, e.g.:

``` C#
                entity.Property(e => e.DepartmentId)
                    .IsRequired()
                    .HasColumnName("DepartmentID");

                entity.Property(e => e.EmploymentStatusId)
                    .IsRequired()
                    .HasColumnName("EmploymentStatusID");

...
                entity.Property(e => e.ProjectId)
                    .IsRequired()
                    .HasColumnName("ProjectID");
``` 

These override the nullability of the FK property itself, making all the relationships required which explains why INNER JOIN is used. You have to remove those calls to `IsRequired()` if you want LEFT OUTER JOIN to be generated.

**Comment by MontyGvMC** [comment]
Thank you for your support.

Well, those code is autogenerated by scaffolding the DB (since we use DB first approach for our project).

I will test it out and report back if this solved my problem.

**Comment by MontyGvMC** [comment]
Without `.IsRequired()` it works as expected. Thank you very much! I will check my complete DB for this calls.

Is the code generator intended to create those `.IsRequired()` calls when scaffolding the DB?

**Comment by divega** [comment]
> Is the code generator intended to create those `.IsRequired()` calls when scaffolding the DB?

No, it is not. That is a bug in `DbContext` scaffolding. But it is already fixed in 2.0.

**Comment by Cubelaster** [comment]
I've got a similar problem. Well, it arises from the same situation, just materializes a bit different. 
Anyway, I have a Soft Delete setup and a GlobalFilter setup to only query for Active rows. 
So, anyways, my Parent class in a relation is SoftDeleted and I guess my EF Core generates an Inner Join but fails to actually join these rows because of the global filter? Am I on the right track?
